src/include/buffer.h

```diff
@@ -1286,6 +1286,23 @@ inline bufferhash& operator<<(bufferhash& l, const bufferlist &r) {
   return l;
 }
 
+static inline
+void copy_bufferlist_to_iovec(const struct iovec *iov, unsigned iovcnt,
+                              bufferlist *bl, int64_t r)
+{
+  auto iter = bl->cbegin();
+  for (unsigned j = 0, resid = r; j < iovcnt && resid > 0; j++) {
+         /*
+          * This piece of code aims to handle the case that bufferlist
+          * does not have enough data to fill in the iov
+          */
+         const auto round_size = std::min<unsigned>(resid, iov[j].iov_len);
+         iter.copy(round_size, reinterpret_cast<char*>(iov[j].iov_base));
+         resid -= round_size;
+         /* iter is self-updating */
+  }
+}
```

Comment:
Instead of copying out of the bufferlist contents into the iovec buffers, could we use a provided iovec to make a bufferlist of static buffers pointing at the iovec buffers and supplying that bufferlist as the target of a read?

Reply:
That sounds like it would be better. This is definitely an area where I was bumping into my lack of understanding so any suggestions on how to change it would be helpful.

Reply:
Look at `buffer::create_static` It will make a `buffer::ptr` referencing an already-allocated region of memory.

Then you can shove it into a buffer list with with `push_back`. In principle you should be able to loop over the iovecs and make a bufferlist that will read/write from the same memory.

Reply:
Delving into the code, eventually the bufferlist is utilized by ObjectCacher::_readx which clears (and disposes) of the buffers already in the bufferlist and attaches new buffers. I think what might be going on is a result of data caching. If ceph is maintaining a data cache, and Ganesha wants to get the data into an iovec, a copy must happen.

I think the solution to eliminating this copy is for the API to change so Ganesha can receive buffers from the filesystem and return them when the RPC I/O is complete. Of course in this case, the buffer would be refcounted, so it would remain in the cache, and multiple references could be passed to multiple Ganesha READ requests.

Doing so of course will obligate Ganesha to not change the data (so we could not encrypt in place for example).

Given what it looks like is going on here, I propose we leave this code alone for now, and have a second pass at async/nonblocking I/O where we do work both in Ganesha and cephfs to remove data copying everywhere possible from end to end.

Reply:
just to be clear, we're just agreeing not to attempt copy elision in this PR?

Reply:
Yes, we will work on removing copies in a separate effort. I think we
actually already have github issues open for eliminating copies.